Source/engine/game_assets.cpp

```diff
@@ -0,0 +1,76 @@
+#include "engine/game_assets.hpp"
+
+#include <algorithm>
+#include <cstdint>
+#include <cstring>
+
+#include "init.h"
+#include "utils/file_util.h"
+#include "utils/log.hpp"
+#include "utils/mpq_sdl_rwops.hpp"
+#include "utils/paths.h"
+
+namespace devilution {
+
+namespace {
+
+bool OpenMpqFile(const char *filename, MpqArchive **archive, uint32_t *fileNumber)
+{
+	const MpqArchive::FileHash fileHash = MpqArchive::CalculateFileHash(filename);
+	const auto at = [=](std::optional<MpqArchive> &src) -> bool {
+		if (src && src->GetFileNumber(fileHash, *fileNumber)) {
+			*archive = &(*src);
+			return true;
+		}
+		return false;
+	};
+
+	return at(font_mpq) || at(lang_mpq) || at(devilutionx_mpq)
+	    || (gbIsHellfire && (at(hfvoice_mpq) || at(hfmusic_mpq) || at(hfbarb_mpq) || at(hfbard_mpq) || at(hfmonk_mpq) || at(hellfire_mpq))) || at(spawn_mpq) || at(diabdat_mpq);

```

Comment:
I think it would be best if we could instead have an ordered map rather then hardcode the order, and set the priority with `LoadMPQ("spawn.mpq", 1000)`, which appears to be how things where originally handled.

This is the original priority values:
diabdat.mpq/spawn.mpq 1000
hellfire.mpq 8000
hfmonk.mpq 8100
hfbard.mpq 8110
hfbarb.mpq 8120
hfmusic.mpq 8200
hfvoice.mpq 8500

(high number are checked fierst)

Suggested values for  our MPQ files
lang_mpq 10200
font.mpq 10100
devilutionx_mpq 10000

Also I think we should add a check for user.mpq to allow for custom content (previously people would usually name it patch_rt.mpq, or use the two unused Hellfire names hfopt1.mpq hfopt2.mpq)

I'm ok with this not being done as part of this PR, just thought it was a good time to bring it up :)

Reply:
That could be an option, although I'm not sure priorities are better than a simple list but is out-of-scope for this PR

Reply:
The benefit of priorities is that it allows us to load some mpq's early (fonts), but prioritize them at a different order then they where loaded.

Reply:
Sure but that could still just be a list. The drawback of priorities is that their effects are non-local.